tao_compiler/mlir/disc/transforms/mhlo_decomp_rewriters.cc

```diff
@@ -132,6 +132,104 @@ LogicalResult SliceOpConvert::matchAndRewrite(mhlo::SliceOp op,
       op, op.getType(), operand, startIndices, limitIndices, strides);
   return success();
 }
+
+struct ScatterOpConverter : public OpRewritePattern<mhlo::ScatterOp> {
+ public:
+  using OpRewritePattern<mhlo::ScatterOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(mhlo::ScatterOp hloOp,
+                                PatternRewriter& rewriter) const override {
+    Operation* op = hloOp.getOperation();
+    auto operands = hloOp.getOperands();
+
+    if (!operands[0]
+             .getType()
+             .cast<RankedTensorType>()
+             .getElementType()
+             .isBF16()) {
+      return failure();
+    }
+
+    auto& body = hloOp.getUpdateComputation();
+    Operation* calc_op = nullptr;
+    body.walk([&](Operation* op) {
+      if (llvm::isa<mhlo::AddOp, mhlo::MulOp, mhlo::MaxOp, mhlo::MinOp>(op))
+        calc_op = op;
+    });
+    if (!calc_op) {
+      return failure();
+    }
+
+    // Insert Type Conversion
+    auto self_bf16_type = operands[0].getType().cast<RankedTensorType>();
+    auto self_fp32_type = RankedTensorType::get(self_bf16_type.getShape().vec(),
+                                                rewriter.getF32Type());
+
+    auto src_bf16_type = operands[2].getType().cast<RankedTensorType>();
+    auto src_fp32_type = RankedTensorType::get(src_bf16_type.getShape().vec(),
+                                               rewriter.getF32Type());
+
+    auto self_convert_type = rewriter.create<mhlo::ConvertOp>(
+        hloOp.getLoc(), self_fp32_type, hloOp.getOperand(0));
+    auto src_convert_type = rewriter.create<mhlo::ConvertOp>(
+        hloOp.getLoc(), src_fp32_type, hloOp.getOperand(2));
+
+    SmallVector<Value> new_operands;
+    new_operands.push_back(self_convert_type->getResult(0));
+    new_operands.push_back(operands[1]);
+    new_operands.push_back(src_convert_type->getResult(0));
+
+    auto scatter_op = rewriter.create<mhlo::ScatterOp>(
+        hloOp.getLoc(), self_fp32_type, new_operands, hloOp->getAttrs());
+
+    auto res_convert_back = rewriter.create<mhlo::ConvertOp>(
+        hloOp.getLoc(), self_bf16_type, scatter_op->getResult(0));
+
+    // Construct updateComputation region, here we treat it as update operation
+    Block& block = scatter_op.getUpdateComputation().emplaceBlock();
+    auto blockArg1Type =
+        RankedTensorType::get({}, self_fp32_type.getElementType());
+    auto blockArg2Type =
+        RankedTensorType::get({}, src_fp32_type.getElementType());
+    block.addArgument(blockArg1Type, hloOp.getLoc());
+    block.addArgument(blockArg2Type, hloOp.getLoc());
+    {
+      OpBuilder::InsertionGuard guard(rewriter);
+      rewriter.setInsertionPointToStart(&block);
+
+      Value retValue;
+      if (llvm::isa<mhlo::AddOp>(calc_op)) {
```

Comment:
can we polish this code block using a lambda function `createOp`?
``` cpp
auto createOp = [&](auto opType) {
    return rewriter.create<decltype(opType)>(op->getLoc(), block.getArgument(0), block.getArgument(1)).getResult();
};

if (llvm::isa<mhlo::AddOp>(calc_op)) {
    retValue = createOp(mhlo::AddOp());
}
```

Reply:
Fixed.